scripts/mayachemtools/docs/modules/man3/DBUtil.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DBUtil 1"
.TH DBUtil 1 "2018-05-15" "perl v5.22.4" "MayaChemTools"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
DBUtil
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
use DBUtil;
.PP
use DBUtil qw(:all);
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBDBUtil\fR module provides the following functions:
.PP
DBConnect, DBDisconnect, DBFetchSchemaTableNames, DBSQLToTextFile,
DBSetupDescribeSQL, DBSetupSelectSQL
.PP
DBUtil package uses Perl \s-1DBI\s0 for interacting with MySQL Oracle, and PostgreSQL
databases.
.SH "FUNCTIONS"
.IX Header "FUNCTIONS"
.IP "\fBDBConnect\fR" 4
.IX Item "DBConnect"
.Vb 1
\&    $DBHandle = DBConnect($DBDriver, $DBName, $DBHost, $DBUser, $DBPassword);
.Ve
.Sp
Connects to a database using specified parameters and returns a \fBDBHandle\fR.
.IP "\fBDBDisconnect\fR" 4
.IX Item "DBDisconnect"
.Vb 1
\&    DBDisconnect($DBHandle);
.Ve
.Sp
Disconnects from a database specified by \fIDBHandle\fR.
.IP "\fBDBFetchSchemaTableNames\fR" 4
.IX Item "DBFetchSchemaTableNames"
.Vb 2
\&    @SchemaTableNames = DBFetchSchemaTableNames($DBDriver, $DBHandle,
\&                       $SchemaName);
.Ve
.Sp
Returns an array of all the table names in a database \fISchemaName\fR.
.IP "\fBDBSetupDescribeSQL\fR" 4
.IX Item "DBSetupDescribeSQL"
.Vb 1
\&    $DescribeSQL = DBSetupDescribeSQL($DBDriver, $TableName, [$SchemaName]);
.Ve
.Sp
Sets up and returns a \s-1SQL\s0 statement to describe a table for MySQ, Oracle or PostgreSQL.
.IP "\fBDBSetupSelectSQL\fR" 4
.IX Item "DBSetupSelectSQL"
.Vb 1
\&    $SelectSQL = DBSetupSelectSQL($DBDriver, $TableName, $SchemaName);
.Ve
.Sp
Sets up and returns a \s-1SQL\s0 statement to retrieve all columns from a table for MySQL,
Oracle, or PostgreSQL.
.IP "\fBDBSQLToTextFile\fR" 4
.IX Item "DBSQLToTextFile"
.Vb 3
\&    $Status = DBSQLToTextFile($DBHandle, $SQL, \e*TEXTFILE, $OutDelim,
\&              $OutQuote, [$ExportDataLabels, $ExportLOBs,
\&              $ReplaceNullStr]);
.Ve
.Sp
Executes a \fI\s-1SQL\s0\fR statement and export all data into a text file.
.SH "AUTHOR"
.IX Header "AUTHOR"
Manish Sud <msud@san.rr.com>
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (C) 2018 Manish Sud. All rights reserved.
.PP
This file is part of MayaChemTools.
.PP
MayaChemTools is free software; you can redistribute it and/or modify it under
the terms of the \s-1GNU\s0 Lesser General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option)
any later version.


mayachemtools/lib/DBUtil.pm
package DBUtil;
#
# File: DBUtil.pm
# Author: Manish Sud <msud@san.rr.com>
#
# Copyright (C) 2018 Manish Sud. All rights reserved.
#
# This file is part of MayaChemTools.
#
# MayaChemTools is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation; either version 3 of the License, or (at your option) any
# later version.
#
# MayaChemTools is distributed in the hope that it will be useful, but without
# any warranty; without even the implied warranty of merchantability of fitness
# for a particular purpose.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with MayaChemTools; if not, see <http://www.gnu.org/licenses/> or
# write to the Free Software Foundation Inc., 59 Temple Place, Suite 330,
# Boston, MA, 02111-1307, USA.
#

use strict;
use Exporter;
use Carp;
use DBI;
use TextUtil;

use vars qw(@ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);

@ISA = qw(Exporter);
@EXPORT = qw(DBConnect DBDisconnect DBFetchSchemaTableNames DBSetupDescribeSQL DBSetupSelectSQL DBSQLToTextFile);
@EXPORT_OK = qw();
%EXPORT_TAGS = (all  => [@EXPORT, @EXPORT_OK]);

# Connect to a specified database...
sub DBConnect {
  my($DBDriver, $DBName, $DBHost, $DBUser, $DBPassword) = @_;
  my($DBHandle, $DataSource);

  if ($DBDriver eq "Oracle") {
    $DataSource = qq(DBI:$DBDriver:$DBHost);
  }
  else {
    $DataSource = qq(DBI:$DBDriver:database=$DBName);
    if ($DBHost) {
      $DataSource .= qq(;host=$DBHost);
    }
  }

  # Don't raise the error; otherwise, DBI functions termiates on encountering an error.
  # All terminations decisions are made outside of DBI functions...
  $DBHandle = DBI->connect($DataSource, $DBUser, $DBPassword, { RaiseError => 0, AutoCommit => 0 }) or croak "Couldn't connect to database...";

  return $DBHandle;
}

# Disconnect from a database...
sub DBDisconnect {
  my($DBHandle) = @_;

  $DBHandle->disconnect or carp "Couldn't disconnect from a database...";
}

# Fetch all table name for a database schema...
sub DBFetchSchemaTableNames {
  my($DBDriver, $DBHandle, $SchemaName) = @_;
  my(@SchemaTableNames, $SQL, $SQLHandle);

  @SchemaTableNames = ();

  $SchemaName = (defined $SchemaName && length $SchemaName) ? $SchemaName : "";

  if ($DBDriver eq "mysql") {
    # Switch schemas...
    $SQL = qq(USE $SchemaName);
    $SQLHandle = $DBHandle->prepare($SQL) or return @SchemaTableNames;
    $SQLHandle->execute or return @SchemaTableNames;
    $SQLHandle->finish or return @SchemaTableNames;

    # Setup to fetch table names...
    $SQL = qq(SHOW TABLES);
  }
  elsif ($DBDriver eq "Oracle") {
    $SQL = qq(SELECT SEGMENT_NAME FROM DBA_SEGMENTS WHERE OWNER = '$SchemaName' AND SEGMENT_TYPE = 'TABLE' ORDER BY SEGMENT_NAME);
  }
  elsif ($DBDriver =~ /^(Pg|Postgres)$/i) {
    $SQL = qq(SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA = '$SchemaName');
  }
  $SQLHandle = $DBHandle->prepare($SQL) or return @SchemaTableNames;
  $SQLHandle->execute or return @SchemaTableNames;

  my(@RowValues, $TableName);
  while (@RowValues = $SQLHandle->fetchrow_array) {
    $TableName = ($DBDriver =~ /^(mysql|Oracle)$/i) ? uc($RowValues[0]) : $RowValues[0];
    if (defined $TableName && length $TableName) {
      push @SchemaTableNames, $TableName;
    }
  }
  $SQLHandle->finish or return @SchemaTableNames;

  return @SchemaTableNames;
}

# Setup describe SQL statement...
sub DBSetupDescribeSQL {
  my($DBDriver, $TableName, $SchemaName);
  my($DescribeSQL);

  $DBDriver = ""; $TableName = ""; $SchemaName = "";
  if (@_ == 3) {
    ($DBDriver, $TableName, $SchemaName) = @_;
  }
  else {
    ($DBDriver, $TableName) = @_;
  }
  $TableName = (defined $TableName && length $TableName) ? $TableName : "";
  $SchemaName = (defined $SchemaName && length $SchemaName) ? $SchemaName : "";

  $DescribeSQL = ($SchemaName) ? ("DESCRIBE " . "$SchemaName" . ".$TableName") : "DESCRIBE $TableName";

  if ($DBDriver eq "Oracle") {
    $DescribeSQL = qq(SELECT COLUMN_NAME "Column_Name", DECODE(NULLABLE, 'N','Not Null','Y','Null') "Null", DATA_TYPE "Data_Type", DATA_LENGTH "Data_Length", DATA_PRECISION "Data_Precision" FROM DBA_TAB_COLUMNS WHERE TABLE_NAME = '$TableName');
    if ($SchemaName) {
      $DescribeSQL .= qq( AND OWNER = '$SchemaName');
    }
    $DescribeSQL .= qq( ORDER BY COLUMN_ID);
  }
  elsif ($DBDriver =~ /^(Pg|Postgres)$/i) {
    $DescribeSQL = qq(SELECT COLUMN_NAME "Column_Name", data_type "Data_Type" FROM information_schema.columns WHERE table_name ='$TableName');
    if ($SchemaName) {
      $DescribeSQL .= " and table_schema = '$SchemaName'";
    }
  }

  return $DescribeSQL;
}

# Setup describe SQL statement...
sub DBSetupSelectSQL {
  my($DBDriver, $TableName, $SchemaName);
  my($SelectSQL);

  $DBDriver = ""; $TableName = ""; $SchemaName = "";
  if (@_ == 3) {
    ($DBDriver, $TableName, $SchemaName) = @_;
  }
  else {
    ($DBDriver, $TableName) = @_;
  }
  $TableName = (defined $TableName && length $TableName) ? $TableName : "";
  $SchemaName = (defined $SchemaName && length $SchemaName) ? $SchemaName : "";

  $SelectSQL = ($SchemaName) ? ("SELECT * FROM " . "$SchemaName" . ".$TableName") : "SELECT * FROM $TableName";

  return $SelectSQL;
}

# Prepare and execute a SQL statement and write out results into
# a text file.
sub DBSQLToTextFile {
  my($DBHandle, $SQL, $TextFile, $OutDelim, $OutQuote, $ExportDataLabels, $ExportLOBs, $ReplaceNullStr);
  my($SQLHandle, $Status);

  $Status = 1;
  $ExportDataLabels = 1;
  $ExportLOBs = 0;
  $ReplaceNullStr = "";
  if (@_ == 8) {
    ($DBHandle, $SQL, $TextFile, $OutDelim, $OutQuote, $ExportDataLabels, $ExportLOBs, $ReplaceNullStr) = @_;
  }
  elsif (@_ == 7) {
    ($DBHandle, $SQL, $TextFile, $OutDelim, $OutQuote, $ExportDataLabels, $ExportLOBs) = @_;
  }
  elsif (@_ == 6) {
    ($DBHandle, $SQL, $TextFile, $OutDelim, $OutQuote, $ExportDataLabels) = @_;
  }
  else {
    ($DBHandle, $SQL, $TextFile, $OutDelim, $OutQuote) = @_;
  }

  # Execute SQL statement...
  $SQLHandle = $DBHandle->prepare($SQL) or return $Status;
  $SQLHandle->execute() or return $Status;

  my($FieldsNum, @FieldNames, @RowValues, @ColNumsToExport, @ColLabels, $ColNum, $ColLabelsLine, @Values, $Value, $ValuesLine);

  $Status = 0;
  # Figure out which column numbers need to be exported...
  $FieldsNum = $SQLHandle->{NUM_OF_FIELDS};
  @FieldNames = @{$SQLHandle->{NAME}};
  @ColNumsToExport = ();
  if ($ExportLOBs) {
    @ColNumsToExport = (0 .. $#FieldNames);
  }
  else {
    my(@FieldTypes, @FieldTypeNames, $Type, $TypeName);
    @FieldTypes = @{$SQLHandle->{TYPE}};
    @FieldTypeNames = map { scalar $DBHandle->type_info($_)->{TYPE_NAME} } @FieldTypes;
    for $ColNum (0 .. $#FieldNames) {
      if ($FieldTypeNames[$ColNum] !~ /lob|bytea/i ) {
	push @ColNumsToExport, $ColNum;
      }
    }
  }

  if ($ExportDataLabels) {
    # Print out column labels...
    @ColLabels = ();
    for $ColNum (@ColNumsToExport) {
      push @ColLabels, $FieldNames[$ColNum];
    }
    $ColLabelsLine = JoinWords(\@ColLabels, $OutDelim, $OutQuote);
    print $TextFile "$ColLabelsLine\n";
  }
  # Print out row values...
  while (@RowValues = $SQLHandle->fetchrow_array) {
    @Values = ();
    for $ColNum (@ColNumsToExport) {
      if (defined($RowValues[$ColNum]) && length($RowValues[$ColNum])) {
	$Value = $RowValues[$ColNum];
      }
      else {
	$Value = $ReplaceNullStr ? $ReplaceNullStr : "";
      }
      push @Values, $Value;
    }
    $ValuesLine = JoinWords(\@Values, $OutDelim, $OutQuote);
    print $TextFile "$ValuesLine\n";
  }
  $SQLHandle->finish or return $Status;
  $Status = 0;

  return $Status;
}

1;

__END__

=head1 NAME

DBUtil

=head1 SYNOPSIS

use DBUtil;

use DBUtil qw(:all);

=head1 DESCRIPTION

B<DBUtil> module provides the following functions:

DBConnect, DBDisconnect, DBFetchSchemaTableNames, DBSQLToTextFile,
DBSetupDescribeSQL, DBSetupSelectSQL

DBUtil package uses Perl DBI for interacting with MySQL Oracle, and PostgreSQL
databases.

=head1 FUNCTIONS

=over 4

=item B<DBConnect>

    $DBHandle = DBConnect($DBDriver, $DBName, $DBHost, $DBUser, $DBPassword);

Connects to a database using specified parameters and returns a B<DBHandle>.

=item B<DBDisconnect>

    DBDisconnect($DBHandle);

Disconnects from a database specified by I<DBHandle>.

=item B<DBFetchSchemaTableNames>

    @SchemaTableNames = DBFetchSchemaTableNames($DBDriver, $DBHandle,
                       $SchemaName);

Returns an array of all the table names in a database I<SchemaName>.

=item B<DBSetupDescribeSQL>

    $DescribeSQL = DBSetupDescribeSQL($DBDriver, $TableName, [$SchemaName]);

Sets up and returns a SQL statement to describe a table for MySQ, Oracle or PostgreSQL.

=item B<DBSetupSelectSQL>

    $SelectSQL = DBSetupSelectSQL($DBDriver, $TableName, $SchemaName);

Sets up and returns a SQL statement to retrieve all columns from a table for MySQL,
Oracle, or PostgreSQL.

=item B<DBSQLToTextFile>

    $Status = DBSQLToTextFile($DBHandle, $SQL, \*TEXTFILE, $OutDelim,
              $OutQuote, [$ExportDataLabels, $ExportLOBs,
              $ReplaceNullStr]);

Executes a I<SQL> statement and export all data into a text file.

=back

=head1 AUTHOR

Manish Sud <msud@san.rr.com>

=head1 COPYRIGHT

Copyright (C) 2018 Manish Sud. All rights reserved.

This file is part of MayaChemTools.

MayaChemTools is free software; you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option)
any later version.

=cut
